### DISCUSSION TITLE
Wait for result of modal screen

### DISCUSSION THREAD

**Discussion by ghost** [Textualize/textual#2559] [Q&A] [answered]
How do you wait for the result of a modal screen before continuing?

E.g. the following code does not wait:
````
    async def action_reschedule_all(self):
        """Reschedule all Jobs"""
        if not self.app.is_logged_in:
            await self.app.push_screen("login")
        self.log("################")
```

**Comment by davep** [comment, marked as answer]
Textual has a way to pass a callback into a screen, and have the screen return the result. It's a fairly recent addition which we've been testing out and so [it still needs to be documented](https://github.com/Textualize/textual/issues/2467)  -- you can see an example [in the PR that added the feature](https://github.com/Textualize/textual/pull/2321).

To see it in action in a real application, see [the `YesNo` dialog in Frogmouth](https://github.com/Textualize/frogmouth/blob/main/frogmouth/dialogs/yes_no_dialog.py ) and then [see it being used to add a yes/no confirmation around deleting something](https://github.com/Textualize/frogmouth/blob/main/frogmouth/widgets/navigation_panes/bookmarks.py#L127-L147) (in this case a bookmark).

**Comment by ghost** [comment]
Thanks for the answer, was able to work things out thanks to it!

**Comment by mzebrak** [comment]
@davep Could you tell me if there are any plans for this feature? I mean not passing a callback to the modal screen but instead just awaiting its result or somehow blocking the execution of further code until the screen closes (asyncio.Event?, asyncio.Lock?).

I have some business logic separated into a `Command` pattern. This is to be reusable in TUI as well as external CLI. There are some commands that require the user to log in, so such a command takes a callback to the login functionality. In the case of a TUI, it corresponds with pushing a login screen. At this point I don't see any possibility to return the result from the command when it is called like:



```python
SomeCommand.register_activate_callback(lambda: self.app.push_screen(LoginScreen()))
some_command = SomeCommand()

result = some_command.execute()  # This is not possible since inside the execute method, LoginScreen is pushed and we can't wait till it closes (also probably that should be `await some_command.execute()`
some_further_logic(result)
```

and instead, we handle this by implementing the `Observer` pattern so someone could subscribe to the result, but it complicates the simple execution of such a command.

```python
SomeCommand.register_activate_callback(lambda: self.app.push_screen(LoginScreen()))
some_command = SomeCommand()

def __on_result(result, exception):
    # We need to react to the result here
    some_further_logic(result)
    pass

some_command.observe_result()
```


Maybe it's not a big inconvenience at the moment, but I think that waiting for the result would improve a lot. I don't think working with lots of callbacks is great. I am waiting for your reply to know what you think.

**Reply by davep** [reply]
> Could you tell me if there are any plans for this feature? I mean not passing a callback to the modal screen but instead just awaiting its result or somehow blocking the execution of further code until the screen closes

None that I'm aware of.
